src/python/pybind11/src/pyxrt.cpp

```diff
@@ -40,126 +43,176 @@ m.attr("XCL_BO_FLAGS_NONE") = py::int_(XCL_BO_FLAGS_NONE);
+                             try
+                             { r.set_arg(i, item.cast<xrt::bo>()); }
+                             catch (std::exception e) {  }
+
+                             try
+                             { r.set_arg<int>(i, item.cast<int>()); }
+                             catch (std::exception e) {  }
 
-py::class_<xrt::kernel>(m, "kernel")
-    .def(py::init([](xrt::device d, const py::array_t<unsigned char> u, const std::string & n, bool e){
-        return new xrt::kernel(d, (const unsigned char*) u.request().ptr, n, e);
-	}),	
-	py::arg(""),
-	py::arg(""),
-	py::arg(""),	
-	py::arg("exclusive")=false)
-  .def("__call__", [](xrt::kernel & k, py::args args) -> xrt::run {    
-	int i =0;
-	xrt::run r(k);
-	
-	for (auto item : args) {
-	  try 
-	    { r.set_arg(i, item.cast<xrt::bo>()); }
-	  catch (std::exception e) {  }
-	  
-	  try 
-	    { r.set_arg<int>(i, item.cast<int>()); }
-	  catch (std::exception e) {  }
-	  
-	  i++;
-	}
-	
-	r.start();
-	return r;
-    })
+                             i++;
+                         }
+
+                         r.start();
+                         return r;
+                     })
     .def("group_id", &xrt::kernel::group_id)
     .def("write_register", &xrt::kernel::write_register)
-    .def("read_register", &xrt::kernel::read_register)
-    ;
+    .def("read_register", &xrt::kernel::read_register);
 
 
 /*
  *
- * XRT:: BO
+ * xrt::bo
  *
  */
-py::class_<xrt::bo>(m, "bo")
-    .def(py::init<xrt::device,size_t,xrt::buffer_flags,xrt::memory_group>())
+py::class_<xrt::bo> pybo(m, "bo");
+
+py::enum_<xrt::bo::flags>(pybo, "flags")
+    .value("normal", xrt::bo::flags::normal)
+    .value("cacheable", xrt::bo::flags::cacheable)
+    .value("device_only", xrt::bo::flags::device_only)
+    .value("host_only", xrt::bo::flags::host_only)
+    .value("p2p", xrt::bo::flags::p2p)
+    .value("svm", xrt::bo::flags::svm)
+    .export_values();
+
+pybo.def(py::init<xrt::device, size_t, xrt::bo::flags, xrt::memory_group>())
     .def(py::init<xrt::bo, size_t, size_t>())
     .def("write", ([](xrt::bo &b, py::buffer pyb, size_t seek)  {
-	  py::buffer_info info = pyb.request();
-	  b.write(info.ptr, info.itemsize * info.size , seek);
-    }))
+                       py::buffer_info info = pyb.request();
+                       b.write(info.ptr, info.itemsize * info.size , seek);
+                   }))
     .def("read", ([](xrt::bo &b, size_t size, size_t skip) {
-	  int nitems = size/sizeof(int);
-	  py::array_t<int> result = py::array_t<int>(nitems);
-	  
-	  py::buffer_info bufinfo = result.request();
-	  int* bufptr = (int*) bufinfo.ptr;
-	  b.read(bufptr, size, skip);
-	  return result;
-     }))
-  .def("sync", ([](xrt::bo &b, xclBOSyncDirection dir, size_t size, size_t offset)  {	
-	b.sync(dir, size, offset);
-      }))
-  .def("size", &xrt::bo::size)
-  .def("address", &xrt::bo::address)
+                      int nitems = size/sizeof(int);
```

Comment:
While you are here, if you think about using `auto` it shows that it should remain a `std::size_t` or an `std::intptr_t` at least.

Reply:
Removed nitems.